spec/controllers/spree/admin/orders/payments/payments_controller_spec.rb

```diff
@@ -91,6 +91,19 @@
           end
         end
 
+        context "where further action is required" do
+          before do
+            allow_any_instance_of(Spree::Payment).to receive(:authorize!) do |payment|
```

Comment:
we need to watch out with `allow_any_instance_of` and avoid as much as possible. It's a known bad pattern that leads to brittle tests discouraged even by the RSpec team :point_right: https://relishapp.com/rspec/rspec-mocks/docs/working-with-legacy-code/any-instance.

It's more verbose but safer to stub `.find` and return the mock instance with mocked methods we can control.

Reply:
Definitely noted that `allow_any_instance_of` is not advised. In this case, the instance of `Spree::Payment` doesn't exist until the test runs (when we call `spree_post`) which is why we do it in multiple specs, I believe. So I'm not sure that stubbing `.find` would work. 

The other approach would be to take the `order` object, which does already exist, and stub out `payments` to return `[]` and then stub out `build` on that in order to return the `payment`. This also doesn't feel that great either since it's diving into the controllers `create` method.

If this most recent build passes I would say that we should take it on separately to look at how we can remove all the instances of `allow_any_instance_of`.

Reply:
Hmm... I think stubbing an association like `order.payments` with an array can be problematic as well, it would normally return an `ActiveRecord::Association` object, which will not always behave in the same way as an `Array`, which can cause broken tests or (silently) inaccurate results. So we would be avoiding one problem by creating another in that case...

Reply:
It may be a legit usage of `allow_any_instance_of` but only if it helps us improve this later on. The sole use of it is a symptom of a design that's hard to test IMO.

Reply:
I added https://github.com/openfoodfoundation/openfoodnetwork/issues/6645 and will mark this as resolved; feel free to reopen.